pkg/signature/kms/aws/signer.go

```diff
@@ -0,0 +1,236 @@
+// when the Signer was created.
+//
+// SignMessage recognizes the following Options listed in order of preference:
+//
+// - WithContext()
+//
+// - WithDigest()
+//
+// - WithCryptoSignerOpts()
+//
+// All other options are ignored if specified.
+func (a *SignerVerifier) SignMessage(message io.Reader, opts ...signature.SignOption) ([]byte, error) {
+	var digest []byte
+	var err error
+	ctx := context.Background()
+
+	for _, opt := range opts {
+		opt.ApplyContext(&ctx)
+		opt.ApplyDigest(&digest)
+	}
+
+	var signerOpts crypto.SignerOpts
+	signerOpts, err = a.client.getHashFunc(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "getting fetching default hash function")
+	}
+	for _, opt := range opts {
+		opt.ApplyCryptoSignerOpts(&signerOpts)
+	}
+
+	hf := signerOpts.HashFunc()
+
+	if len(digest) == 0 {
+		digest, hf, err = signature.ComputeDigestForSigning(message, hf, awsSupportedHashFuncs, opts...)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return a.client.sign(ctx, digest, hf)
+}
+
+// PublicKey returns the public key that can be used to verify signatures created by
+// this signer. If the caller wishes to specify the context to use to obtain
+// the public key, pass option.WithContext(desiredCtx).
+//
+// All other options are ignored if specified.
+func (a *SignerVerifier) PublicKey(opts ...signature.PublicKeyOption) (crypto.PublicKey, error) {
+	ctx := context.Background()
+	for _, opt := range opts {
+		opt.ApplyContext(&ctx)
+	}
+
+	return a.client.public(ctx)
+}
+
+// VerifySignature verifies the signature for the given message. Unless provided
+// in an option, the digest of the message will be computed using the hash function specified
+// when the SignerVerifier was created.
+//
+// This function returns nil if the verification succeeded, and an error message otherwise.
+//
+// This function recognizes the following Options listed in order of preference:
+//
+// - WithDigest()
+//
+// - WithCryptoSignerOpts()
+//
+// - WithContext()
+//
+// All other options are ignored if specified.
+func (a *SignerVerifier) VerifySignature(sig, message io.Reader, opts ...signature.VerifyOption) (err error) {
+	ctx := context.Background()
+
+	for _, opt := range opts {
+		opt.ApplyContext(&ctx)
+	}
+
+	return a.client.verify(ctx, sig, message, opts...)
+}
+
+// VerifySignatureRemotely verifies the signature for the given message via KMS
+// API call. Unless provided in an option, the digest of the message will be
+// computed using the hash function specified when the SignerVerifier was created.
+//
+// This function returns nil if the verification succeeded, and an error message otherwise.
+//
+// This function recognizes the following Options listed in order of preference:
+//
+// - WithDigest()
+//
+// - WithCryptoSignerOpts()
+//
+// - WithContext()
+//
+// All other options are ignored if specified.
+func (a *SignerVerifier) VerifySignatureRemotely(sig, message io.Reader, opts ...signature.VerifyOption) (err error) {
```

Comment:
I like that you're giving the choice here, but I'm wondering whether we should add a new option `WithRemoteVerification(bool)` that flips between local or remote verification instead of having a different method signature. I could be convinced either way, but just adding the thought for discussion.

Reply:
Good call, I think that would be cleaner.

Reply:
Added that option in latest commit.